### PR TITLE
[Bug] Link preview improvement

### PR DESCRIPTION
**Pull request by maggievu-pdftron**
Addressed these issues
- fix link preview popup style
- add more link exclusions for phone and email (with link that start with #)
- check if link href is absolute (even if href is relative `el.href` will return the absolute path)
- traverse recursively to parent that has overflow hidden to override this behavior to show the popup

### CODE REVIEW

**Comment by maggievu-pdftron** [comment]
[Linear WVM-203](https://linear.app/pdftron/issue/WVM-203/webviewer-html-link-preview)

**Comment by maggievu-pdftron** [comment]
[Linear WVM-240 ](https://linear.app/pdftron/issue/WVM-240/minor-refactoring-in-html-proxy-server)

File: src/assets/constants.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
+/* eslint-disable @typescript-eslint/no-unused-vars */
+/* eslint-disable no-unused-vars */
+const getPageHeight = () => {
```
Comment (khein-pdftron) [inline src/assets/constants.js:3]: This file is weird. `constants` is a weird name for it to contain the `getPageHeight` function. `shared` is a better name. Let's rename it to that.

**Review by khein-pdftron — CHANGES_REQUESTED** [review]

File: src/assets/constants.js (hunk 1 above)
Comment (maggievu-pdftron) [reply src/assets/constants.js]: `constants` is renamed to `shared`.

**Comment by maggievu-pdftron** [comment]
Also needs this [PR from WV-HTML](https://github.com/XodoDocs/WebViewer-html/pull/45)

File: src/assets/linkPreview.js [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+        `;
+
+        elem.appendChild(popupContainer);
+        let undoOverflow;
+        let intersectionobserver;
+        let elemOriginalStyles = [];
+        let undoFunctions = [];
+
+        const traverseToParentWithOverflowHidden = (targetElement) => {
+          const parentElement = targetElement?.parentElement;
```
Comment (khein-pdftron) [inline src/assets/linkPreview.js:142]: This name doesn't seem accurate. Wouldn't `setParentsToOverflowVisible` be more accurate?

File: src/assets/linkPreview.js [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+  try {
+    await navigator.clipboard.writeText(url);
+    button.dataset.title = 'Link Copied';
+  } catch (err) {
+    button.dataset.title = 'Error Copying';
+  }
+};
+
+
+const faviconEmptySpan = () => {
```
Comment (khein-pdftron) [inline src/assets/linkPreview.js:20]: Not used anywhere. Function should be removed?

File: src/assets/linkPreview.js [hunk 4]
```diff
@@ -0,0 +1,250 @@
+/* eslint-disable no-console */
+/* eslint-disable no-undef */
+/* eslint-disable @typescript-eslint/no-unused-vars */
+/* eslint-disable no-unused-vars */
```
Comment (khein-pdftron) [inline src/assets/linkPreview.js:4]: I would prefer to not disable these rules for the whole file and instead manually disable them for specific lines:
`// eslint-disable-next-line no-unused-vars, @typescript-eslint/no-unused-vars`
and
`// eslint-disable-next-line no-undef`

File: src/assets/linkPreview.js [hunk 5]
```diff
@@ -0,0 +1,250 @@
+/* eslint-disable no-console */
```
Comment (khein-pdftron) [inline src/assets/linkPreview.js:1]: Doesn't seem to be needed

File: src/assets/blockNavigation.js [hunk 6]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable no-undef */
```
Comment (khein-pdftron) [inline src/assets/blockNavigation.js:1]: use `// eslint-disable-next-line no-undef`. See other comment.

File: src/assets/shared.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
+/* eslint-disable @typescript-eslint/no-unused-vars */
+/* eslint-disable no-unused-vars */
```
Comment (khein-pdftron) [inline src/assets/shared.js:2]: use `// eslint-disable-next-line specific-rule`. See other comment.

File: src/assets/getTextData.js [hunk 8]
```diff
@@ -1,7 +1,8 @@
+/* eslint-disable no-undef */
```
Comment (khein-pdftron) [inline src/assets/getTextData.js:1]: use `// eslint-disable-next-line specific-rule`. See other comment.

File: src/index.ts [hunk 9]
```diff
@@ -384,11 +412,18 @@ const createServer = ({
...
-            const styleTag = `<style type='text/css' id='pdftron-css'>${blockNavigationStyle}</style>`;
+            const styleTag = `
+              <style type='text/css' id='pdftron-css'>${blockNavigationStyle}</style>
+              <style type='text/css'>${linkPreviewStyle}</style>
+            `;
             const globalVarsScript = `<script type='text/javascript' id='pdftron-js'>window.PDFTron = {}; window.PDFTron.urlToProxy = '${cookiesUrl}';</script>`;
-            const debounceScript = `<script type='text/javascript'>${debounceJS}</script>`;
-            const navigationScript = `<script type='text/javascript'>${blockNavigationScript}</script>`;
-            const textScript = `<script type='text/javascript'>${sendTextDataScript}</script>`;
+            const scriptTag = `
```
Comment (khein-pdftron) [inline src/index.ts:420]: Nitpick, should be called `scriptTags`

**Review by khein-pdftron — CHANGES_REQUESTED** [review]
The popup looks beautiful! And I love the feedback to copying a link. Amazingly done.

Left a few comments.

**Pull request merged**